### PR TITLE
wgpu-native: init at 22.1.0.5

### LINKED ISSUE #344689: Package request: wgpu
**Project description**

Rust implementation of the WebGPU standard, right now there is only a wgpu-utils that doesn't provide any header. This would provide headers to use for any other application.

Note: There is also another implementation (google dawn) of the standard with an open package request issue.

**Metadata**

* homepage URL:https://wgpu.rs/ , https://crates.io/crates/wgpu
* source URL:https://github.com/gfx-rs/wgpu
* license: mit + apache 2
* platforms: request for linux, but it should work for any other OS

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

### PR DESCRIPTION
**Pull request by niklaskorz**
Another try after #345248 and #348084.
Closes #344689.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Review by Aleksanaa — CHANGES_REQUESTED** [review]
Still this problem. https://github.com/NixOS/nixpkgs/pull/348084#issuecomment-2445804206

> > @Aleksanaa are there any other problems? Thanks for the help btw!

> Sorry, I missed this thread because I was busy. I think the main problem is that we don't know how other projects will use wgpu-native, so we don't know how to prepare headers for them. If our method is wrong, it may have the opposite effect. Generally, libraries in Nixpkgs have some programs depending on it, so we know we packaged it correctly. 

I'd like to see at least one consumer of it, so we can ensure that we do things in the expected way.

**Comment by niklaskorz** [comment]
> I'd like to see at least one consumer of it, so we can ensure that we do things in the expected way.

Fair enough, I'll add the wgpu-native examples first for testing and then look for a user-facing application to include

**Comment by niklaskorz** [comment]
Examples are now packaged under `wgpu-native.examples`. Tested on Darwin and NixOS (on a Wayland desktop).

File: pkgs/by-name/wg/wgpu-native/package.nix [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+  meta = {
+    description = "Native WebGPU implementation based on wgpu-core";
+    homepage = "https://github.com/gfx-rs/wgpu-native";
+    license = with lib.licenses; [
+      mit
+      asl20
+    ];
+    maintainers = with lib.maintainers; [ niklaskorz ];
+    platforms = lib.platforms.all;
```
Comment (Aleksanaa) [inline pkgs/by-name/wg/wgpu-native/package.nix:61]:
```suggestion
```

`rustPlatform` sets its available platforms

File: pkgs/by-name/wg/wgpu-native/examples.nix [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+    runHook postInstall
+  '';
+
+  meta = {
+    description = "Examples for the native WebGPU implementation based on wgpu-core";
+    homepage = "https://github.com/gfx-rs/wgpu-native";
+    license = lib.licenses.mit;
+    maintainers = with lib.maintainers; [ niklaskorz ];
+    platforms = lib.platforms.all;
```
Comment (Aleksanaa) [inline pkgs/by-name/wg/wgpu-native/examples.nix:89]:
```suggestion
  meta = wgpu-native.meta // {
    description = "Examples for the native WebGPU implementation based on wgpu-core";
```

**Review by Aleksanaa — APPROVED** [review]
Looks good!

File: pkgs/by-name/wg/wgpu-native/examples.nix (hunk 2 above)
Comment (niklaskorz) [reply pkgs/by-name/wg/wgpu-native/examples.nix]: Thx!

**Comment by Aleksanaa** [comment]
@ofborg build wgpu-native wgpu-native.passthru.examples

**Pull request merged**